### PR TITLE
Create an APB for the automation broker deployment

### PR DESCRIPTION
**Pull request by djzager**

### CODE REVIEW

File: templates/broker-auth.clusterrole.yaml [hunk 1]
```diff
@@ -0,0 +1,42 @@
...
+    verbs: ["create"]
+  - apiGroups: ["authentication.k8s.io"]
+    resources: ["tokenreviews"]
+    verbs: ["create"]
+  - apiGroups: ["networking.k8s.io", ""]
+    # attributeRestrictions: null
+    resources: ["networkpolicies"]
+    verbs: ["create", "delete"]
+  - apiGroups: ["automationbroker.io"]
+    # attributeRestrictions: null
```
Comment (djzager) [inline templates/broker-auth.clusterrole.yaml]: These aren't actually supported in the python-client. What happens if we don't specifically set them to `null`?

File: templates/broker-auth.secret.yaml [hunk 2]
```diff
@@ -0,0 +1,10 @@
+---
+
+apiVersion: v1
+kind: Secret
+metadata:
+  name: {{ broker_name }}-auth-secret
+  namespace: {{ broker_namespace }}
+data:
+  username: {{ broker_basic_auth_username | b64encode }}
```
Comment (djzager) [inline templates/broker-auth.secret.yaml]: I figured @fabianvf would appreciate this.

**Comment by dymurray** [comment]
Ah... my error was probably because I didn't set the cluster var. Retrying.

**Comment by dymurray** [comment]
```
changed: [localhost] => (item=broker.configmap.yaml)
An exception occurred during task execution. To see the full traceback, use -vvv. The error was: TypeError: __init__() got an unexpected keyword argument 'apiGroups'
failed: [localhost] (item=broker-auth.clusterrole.yaml) => {"changed": false, "item": "broker-auth.clusterrole.yaml", "module_stderr": "Traceback (most recent call last):\n  File \"/tmp/ansible_0SKNA_/ansible_module_k8s_raw.py\", line 163, in <module>\n    main()\n  File \"/tmp/ansible_0SKNA_/ansible_module_k8s_raw.py\", line 159, in main\n    KubernetesRawModule().execute_module()\n  File \"/tmp/ansible_0SKNA_/ansible_modlib.zip/ansible/module_utils/k8s/raw.py\", line 142, in execute_module\n  File \"/tmp/ansible_0SKNA_/ansible_modlib.zip/ansible/module_utils/k8s/helper.py\", line 136, in object_from_params\n  File \"/tmp/ansible_0SKNA_/ansible_modlib.zip/ansible/module_utils/k8s/helper.py\", line 293, in __set_obj_attribute\n  File \"/tmp/ansible_0SKNA_/ansible_modlib.zip/ansible/module_utils/k8s/helper.py\", line 483, in __compare_obj_list\nTypeError: __init__() got an unexpected keyword argument 'apiGroups'\n", "module_stdout": "", "msg": "MODULE FAILURE", "rc": 1}
```

**Comment by djzager** [comment]
There are a few issues with k8s that I need to work through. Specifically, creating the tls stuff.

**Comment by djzager** [comment]
@dymurray I think this is cleaned up. Let me know if you have any more issues. @rthallisey  if you could have a look at this from the k8s perspective that would be really helpful.

File: tasks/kubernetes.yaml [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  when: cluster == 'kubernetes' and apb_action == 'provision'
+
+- name: 'Generate OpenSSL cert for broker'
+  command: >
+    openssl req -nodes -x509 -newkey rsa:4096
+    -keyout /tmp/{{ broker_name }}-cert/key.pem
+    -out /tmp/{{ broker_name }}-cert/cert.pem
+    -days 365
+    -subj "/CN={{ broker_name }}.{{ broker_name }}.svc"
+  when: cluster == 'kubernetes' and apb_action == 'provision'
```
Comment (rthallisey) [inline tasks/kubernetes.yaml]: :+1:

File: tasks/kubernetes.yaml [hunk 4]
```diff
@@ -0,0 +1,55 @@
...
+#  openssl_certificate:
+#    path: /tmp/{{ broker_name }}-cert/cert.pem
+#    privatekey_path: /tmp/{{ broker_name }}-cert/key.pem
+#    subject: "/CN={{ broker_name }}.{{ broker_name }}.svc"
+#    provider: selfsigned
+- name: 'Create directory for OpenSSL cert'
+  file:
+    path: /tmp/{{ broker_name }}-cert
+    state: directory
+  when: cluster == 'kubernetes' and apb_action == 'provision'
```
Comment (fabianvf) [inline tasks/kubernetes.yaml]: I think it's cleaner to encapsulate tasks that share a condition in a block or import task. Also prevents you from missing a conditional update somewhere if something changes.

Comment (djzager) [reply tasks/kubernetes.yaml]: Yeah. I can definitely improve this file and the OpenShift file to encapsulate better.

Comment (djzager) [reply tasks/kubernetes.yaml]: I believe I fixed it with the latest commit @fabianvf let me know what you think.

File: templates/broker-auth.clusterrole.yaml (hunk 1 above)
Comment (fabianvf) [reply templates/broker-auth.clusterrole.yaml]: I don't think anything happens

File: tasks/generate_certificate.yaml [hunk 5]
```diff
@@ -0,0 +1,48 @@
...
+        broker_ca_crt: "{{ lookup('file', '/tmp/' + broker_name + '-cert/cert.pem') | b64encode }}"
+        broker_client_secret: "{{
+          lookup(
+            'k8s',
+            kind='ServiceAccount',
+            namespace=broker_namespace,
+            resource_name=broker_name + '-client'
+            ) | json_query('secrets') | first | json_query('name')
+        }}"
+  when: cluster == 'kubernetes'
```
Comment (fabianvf) [inline tasks/generate_certificate.yaml]: not to nitpick, but for large blocks like these I like to put the when at the top (like after name), because the whitespace can be kind of hard to parse.

File: tasks/generate_certificate.yaml [hunk 6]
```diff
@@ -0,0 +1,48 @@
+---
+
+# TODO: Figure out how to use openssl from ansible
+#- name: Generate a Self Signed OpenSSL certificate
+#  openssl_certificate:
+#    path: /tmp/{{ broker_name }}-cert/cert.pem
+#    privatekey_path: /tmp/{{ broker_name }}-cert/key.pem
+#    subject: "/CN={{ broker_name }}.{{ broker_name }}.svc"
+#    provider: selfsigned
+- name: 'Create OpenSSL Cert for Broker'
```
Comment (fabianvf) [inline tasks/generate_certificate.yaml]: just FYI I don't think the task name is displayed in the output for blocks.

File: templates/bundle.crd.yaml [hunk 7]
```diff
@@ -0,0 +1,155 @@
+---
```
Comment (fabianvf) [inline templates/bundle.crd.yaml]: Since this isn't a template it might make more sense to put it in files/

File: tasks/generate_certificate.yaml (hunk 6 above)
Comment (djzager) [reply tasks/generate_certificate.yaml]: That's what I get for trusting the example :sunglasses: http://docs.ansible.com/ansible/latest/user_guide/playbooks_blocks.html#id1
